text/0876-support-for-es-accessor-decorators.md

```diff
@@ -0,0 +1,420 @@
+
+[spec]: https://github.com/tc39/proposal-decorators
+[0408]: https://rfcs.emberjs.com/id/0408-decorators
+[0440]: https://rfcs.emberjs.com/id/0440-decorator-support
+[0724]: https://rfcs.emberjs.com/id/0724-road-to-typescript/
+
+[^other-transpilers]: and other transpilers like SWC and ESBuild, which will be increasingly viable for Ember users as we migrate to Embroider
+
+
+## Detailed design
+
+When we originally [introduced support][0408] for decorators, we [specified][0440] our support policy and approach for the Stage 1 decorators:
+
+> Stage 1 decorators will be considered a first class Ember API. They will be
+supported until:
+> 
+> 1. A new RFC has been made to introduce a new decorator API as a parallel API to the stage 1 decorators.
+> 2. A deprecation RFC has been created for Stage 1 decorators.
+> 
+> They will follow the same deprecation flow, and same SemVer requirements, as any other Ember feature.
+
+This RFC is the first step in that process: it introduces a new decorator API as a parallel API to the stage 1 decorators. A follow-on RFC will separately deprecate the existing Stage 1 decorator APIs.
+
+
+### Background
+
+The current basic form of property decorators looks like this:
+
+```js
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import { service } from '@ember/service';
+
+export default class Example extends Component {
+  @tracked localState;
+  @service session;
+}
+```
+
+The spec changed both the syntax and the semantics of these kinds of decorators. The semantics are easy enough for us to change but **do** require a new implementation (discussed below). The syntax is also a straightforward change, simply requiring the additoin of the new `accessor` keyword:
+
+```js
+export default class Example extends Component {
+  @tracked accessor localState;
+  @service accessor session;
+}
+```
+
+This `accessor` keyword is exactly equivalent to having a `get` and `set` pair with a backing private field (but with a name that is invisible), like this:
+
+```js
+// this:
+class Example {
+  accessor foo;
+}
+
+// compiles to roughly this:
+class Example {
+  #foo_1234asdfLOL;
+
+  get foo() {
+    return this.#foo_1234asdfLOL;
+  }
+
+  set foo(value) {
+    this.#foo_1234asdfLOL = value;
+  }
+}
+```
+
+Decorators can then layer on top of this. This is more or less what `@tracked`, `@service`, etc. already do—but a spec decorator gets a very different argument, and behaves fairly differently than the legacy decorators implementation.
+
+We *could* try to absorb that with a single overload, but that has a much higher risk of bugs *and* doesn’t give us any hooks for linting etc. Linting matters because we want to be able to tell people statically (not just with annoying Babel compile errors) that they *must* use `accessor` with this version instead.
```

Comment:
a lint can read the package.json and see what version of package we'd have that supports this, and then autofix-insert `accessor` -- making the overload more reasonable, imo

Reply:
The problem is not just the consumer side, it's also the *authoring* side. Those overloads are already quite heavy, and adding more complexity there substantially ups the challenge of shipping non-buggy code. It's not impossible, but  it's far easier to just codemod to the new thing and have much simpler lint rule maintenance as well.

It's also not possible to see purely from the `package.json`: we *have* to support both legacy and spec decorators, so "what version are we on?" doesn't answer the question. That lint rule would have to go parse `babel.config.js` and `ember-cli-build.js`, and/or boot Ember CLI to get the resolved result of the config.